README
Version 0.5

TODO
static ip's

KNOWN ISSUES:
Long names in the manual ssid can give trouble

DEPENDENCIES:
dhclient
wireless-tools
wpa_supplicant

INSTALL INSTRUCTIONS
To build just run the following command
$ make
# make install

REMOVE INSTRUCTIONS
To remove the program just run:
# make remove

To purge the configuration files with it run:
# make purge

PROGRAM INSTRUCTIONS
-dwifi
Help and examples can be found when running
$ ./dwifi -h

-dwifid
The daemon dwifid is automatically installed when make install is runned.
It reads the priority list from the file /etc/wifi/.networks
You can start/stop/restart the daemon by running:
# service dwifid {start|stop|restart}
To remove the automatic startup just run the following command:
# update-rc.d dwifid remove
To reverse the removal just run
# update-rc.d dwifid default

CONFIG INSTRUCTIONS
Sample config files are included for the university eduroam network.
There is a config creater built in but this is only for wpa2

WPA2 config files can be easily generated by typing:
# wpa_passphrase ssid password > /etc/wifi/ssid.conf


CHANGELOG
2013-Jun-19
	Version 0.1:
	First version, only manual configs and wpa2 networks.

2013-Jun-23
	Version 0.2:
	Support for unprotected networks.

2013-Jun-26
	Version 0.3:
	Config creater included for wpa2 networks

2013-Aug-29
	Version 0.4:
	Manual ssid command line argument added for future daemon support

2013-Aug-31
	Version 0.4a:
	List only option added for future daemon



src/creator.c
#include "./creator.h"	//File it extends
#include <stdio.h>		//File io
#include <stdlib.h>		//Standard library
#include <string.h>		//Working and printing cstrings

/* Reads a single word from stdin */
char* read_word()
{
	char* wpa = (char*)malloc(4048*sizeof(char));
	int i = 0;
	char c;
	fflush(stdin);
	while((c = getchar()) != '\n') wpa[i++] = c;
	wpa[i] = '\0';
	return wpa;
}

/* Reads a wpa key from stdin */
char* read_wpa()
{
	printf("Please enter the passphrase: ");
	char* wpa;
	while(strlen(wpa = read_word())<8)
		printf("WPA code should be a minimal of 8 characters\nPlease try again: ");
	return wpa;
}

/* Creates a config for the given network */
void create_config(struct wifinet* network, char* configpath)
{
	char command[4048];
	printf("CONFIG CREATION IS STILL EXPERIMENTAL!!!\n");
	wifinet_print(network);
	switch(network->encr)
	{
		case wpa1:
		case wpa2:
			sprintf(command, "wpa_passphrase %s %s > %s/%s.conf", network->ssid, read_wpa(), configpath, network->ssid);
			printf("%s\n", command);
			break;	
	}	
	system(command);
	printf("Config written to %s/%s.conf\n", configpath, network->ssid);
}


src/wifinet.c
#include "./wifinet.h" //The file it extends
#include <stdio.h>	  //File io
#include <string.h>	  //Workinng and printing cstrings
#include <stdlib.h>	  //Standard lib

char wpa1id[] = "WPA Version 1";
char wpa2id[] = "IEEE 802.11i/WPA2 Version 1";

/* Prints a array of wifinetworks */
void wifinets_print(struct wifinet* w, int n)
{
	printf("# strength\tencryption\tssid (bssid)\n");
	int i;
	for(i = 0; i<n; i++)
		wifinet_print(&w[i]);
}

/* Prints a single wifinetwork */
void wifinet_print(struct wifinet* w)
{
	printf("%c. %.1fdbm\t%s\t\t%s (%s)\n", w->str, w->id, w->encr==no ? "open" : w->encr==wpa1 ? "wpa1" : "wpa2", w->ssid, w->hwaddr);
}

/* Reads a wifinet from a file */
struct wifinet* get_wifinets(int* nets, FILE* f)
{
	char* l = malloc(4048*sizeof(char));
	struct wifinet* networks = malloc(36*sizeof(struct wifinet));
	(*nets) = -1;
	while(read_line(f, l))
	{	
		char prev,  cur;
		int n = 0;
		while((cur = l[n++])==' ');		/* Skip all the spaces */
		prev = cur;
		cur = l[n];
		if(prev=='C' && cur=='e')			/* Read hardware adress */
		{
			(*nets)++;
			networks[(*nets)].id = (*nets)<10 ? (*nets)+48 : (*nets)-10+97;
			memcpy(networks[(*nets)].hwaddr, &l[n+18], 17);
		}
		if(prev=='E' && cur=='S') 			/* Read SSID */
		{
			int ssidlength = 0;
			n += 6;
			while(l[n+ssidlength]!='\0')
				ssidlength++;
			memcpy(networks[(*nets)].ssid, &l[n], ssidlength-2);
		}
		if(prev=='Q' && cur=='u') 			/* Read signal strength */
		{
			char b[3];
			memcpy(b, &l[n+27], 3);
			networks[(*nets)].str = atof(b);			
		}
		if(prev=='I' && cur=='E')			/* Read encryption kind */
		{
			char* t = strstr(&l[n+3], wpa1id);
			if(t!=NULL)
			{
				networks[(*nets)].encr = networks[(*nets)].encr==wpa2 ? wpa2 : wpa1;
			}
			if((t = strstr(&l[n+3], wpa2id))!=NULL)
			{
				networks[(*nets)].encr = wpa2;
			}		
		}
	}
	fclose(f);
	return networks;
}


src/dwifi.c
#include <stdio.h>		//File io
#include <stdlib.h>		//Printing to the screen
#include <string.h>		//Working with and printing cstrings
#include "./wifinet.h"	//The wifinet tools
#include "./creator.h"	//Network creator

/*Prototyping the functions */
int read_line(FILE* f, char* l);
int yn_dialog(char* l);
void connect(struct wifinet* network);
void parse_args(int argc, char** argv);

/* The global variables for the software */
char* silent = "";
char list_only = 0;
struct wifinet final; int finale = 0;
char* logfile = ".essid";
char* interface = "wlan0";
char* driver = "wext";
char* configpath = "/etc/wifi";

/* Reading one line from a file */
int read_line(FILE* f, char* l)
{
	char c;
	int i = 0;
	while((c = fgetc(f)) != EOF && c != '\n')
		if(c>=32)
			l[i++] = c;
	l[i+1] = '\0';
	return c!=EOF;	
}

/* Simple yes no dialog with the string as argument */
int yn_dialog(char* l)
{
	fflush(stdin);
	printf("%s [y]es/[n]o? ", l);
	char c;
	while((c = getchar())!='y' && c!='n' && c!='Y' && c!='N')
		fflush(stdin);
	return c=='y' || c=='Y';
}

/* The connect routine */
void connect(struct wifinet* network)
{	
	if(strlen(silent)==0) printf("trying to connect to %s\n", network->ssid);
	char conf[4048];
	sprintf(conf, "%s/%s.conf", configpath, network->ssid); 
	FILE* f;	
	if(f = fopen(conf, "r"))
	{
		fclose(f);
	}
	else if(network->encr!=no)
	{
		printf("Config file not found or in use\nI'm looking for: %s\n", conf);
		if(yn_dialog("Want to create the config for the network?"))
		{
			create_config(network, configpath);
			connect(network);
		}
		else
		{
			exit(EXIT_FAILURE);
		}
	}
	char path[4048];
	system("killall -q wpa_supplicant dhclient");
	sprintf(path, "ifconfig %s down%s", interface, silent);
	system(path);
	sprintf(path, "ifconfig %s up%s", interface, silent);
	system(path);
	sprintf(path, "iwconfig %s essid \"%s\"%s", interface, network->ssid, silent);
	system(path);
	switch(network->encr)
	{
		case no:
			sprintf(path, "echo Warning connecting to a unprotected network!");
			break;
		case wpa1:
		case wpa2:
			sprintf(path, "wpa_supplicant -i%s -c%s -D%s -B%s", interface, conf, driver, silent);
			break;			
	} 
	system(path);
	sprintf(path, "dhclient %s%s", interface, silent);
	system(path);
	sprintf(path, "ifconfig %s%s", interface, silent);
	system(path);
}

/* Command line argument parsing */
void parse_args(int argc, char** argv)
{
	int i;
	for(i = 1; i<argc; i++)
	{
		if(strcmp(argv[i], "-h") == 0)
		{
			printf("Usage:\n\t%s [options]\n", argv[0]);
			printf("Options:\n");
			printf("\t-h        Display this help\n");
			printf("\t-l        Only list the currently available ssids\n");
			printf("\t-s        To make the program silent(pipe to /dev/null\n");			
			printf("\t-i x      where x is the interface(default: wlan0)\n");
			printf("\t-d x      where x is the driver(default: wext)\n");
			printf("\t-c x      where x is the config root path(default: /etc/wifi)\n");
			printf("\t-ssid x y where x is the essid and y the encryption type(open/wpa2)\n");
			printf("\nThe config file has to match the essid for example if you want to connect to ");
			printf("the \"eduroam\" network you have to have a config file /etc/wifi/eduroam.conf\n");
			printf("The config file for a simple WPA2 network can be easily created by wpa_passphrase ");
			printf("for example:  \n");
			printf("wpa_passphrase essid password >> /etc/wifi/eduroam.conf\"\n");
			exit(EXIT_SUCCESS);
		}
		if(strcmp(argv[i], "-i") == 0 && argc>i)
		{
			interface = (char*)malloc(50*sizeof(char));
			strcpy(interface, argv[i+1]);
			if(strlen(silent)==0) printf("Using manual interface: %s\n", interface);
		}
		if(strcmp(argv[i], "-d") == 0 && argc>i)
		{
			driver = (char*)malloc(50*sizeof(char));
			strcpy(driver, argv[i+1]);
			if(strlen(silent)==0) printf("Using manual driver: %s\n", driver);
		}
		if(strcmp(argv[i], "-c") == 0 && argc>i)
		{
			configpath = (char*)malloc(50*sizeof(char));
			strcpy(configpath, argv[i+1]);
			if(strlen(silent)==0) printf("Using manual configpath: %s\n", configpath);
		}
		if(strcmp(argv[i], "-ssid") == 0 && argc>i)
		{
			strcpy(final.ssid, argv[i+1]);
			char* ssidc;
			char* configprefix;
			configprefix = (char*)malloc(250*sizeof(char));
			strcpy(configprefix, configpath);
			ssidc = (char*)malloc(strlen(final.ssid)*sizeof(char));
			strcpy(ssidc, final.ssid);
			strcat(ssidc, ".conf");
			sprintf(configprefix, "%s/%s", configprefix, ssidc);
			FILE* is;
			if(is = fopen(configprefix, "r"))
			{
				fclose(is);
				final.encr = wpa2;
			}	
			else
			{
				final.encr = no;
			}
			if(strlen(silent)==0) 
			{
				printf("Using manual ssid: %s\n", argv[i+1]);
				printf("Connecting to a %s network\n", final.encr == no ? "open" : "protected");
				printf("Warning, if there is no configuration file present an unprotected network is assumed...\n");
			}
			finale = 1;
		}
		if(strcmp(argv[i], "-s") == 0)
		{
			silent = (char*)malloc(15*sizeof(char));
			strcpy(silent, " >> /dev/null");
		}
		if(strcmp(argv[i], "-l") == 0)
		{
			list_only = 1;
		}
	}
}

/* Main function */
int main(int argc, char* argv[]) 
{
	parse_args(argc, argv);
	if(!finale)
	{
		/* Put the interface up and scan for wireless networks */
		char path[100];
		sprintf(path, "ifconfig %s up%s", interface, silent);
		system(path);
		sprintf(path, "iwlist %s scan > %s", interface, logfile);
		system(path);

		int nlines;
		FILE* f = fopen(logfile, "r");
		struct wifinet* networks = get_wifinets(&nlines, f);
		remove(logfile);	
	
		printf("Done scanning: %i networks found\n", nlines);
		if(nlines==0)
			exit(EXIT_SUCCESS);
		
		wifinets_print(networks, nlines);
		if(list_only==1)
			exit(EXIT_SUCCESS);
	
		char c;
		printf("Please enter character corresponding to the network(X to quit): ");
		while(!(((c = getchar())>=48 && c<48+(nlines>10?10:nlines) ) || (nlines>10 && c>=97 && c<=97+nlines-10)))
			if(c=='X' && yn_dialog("Are you sure? ")) 
				exit(EXIT_SUCCESS);
		
		connect(&networks[c>=48 && c<=57 ? c-48 : c-97+10]);
	}
	else
	{
		connect(&final);
	}
	exit(EXIT_SUCCESS);
}
